app/src/main/java/de/westnordost/streetcomplete/data/osm/osmquests/SpatialCache.kt

```diff
@@ -0,0 +1,140 @@
+import de.westnordost.streetcomplete.util.math.isCompletelyInside
+
+// inspired by Android LruCache
+// fetch needs get data from db and fill other caches, like questKey -> Quest
+// onRemoveTile needs to clean other caches, like questKey -> Quest
+// careful: put does not actually put key into cache if tile is not cached -> update db blocking if key was not put
+// TODO: synchronized stuff? didn't get related crash so far...
+// TODO: tile zoom as parameter? could smaller tiles help for mapDataController?
+class SpatialCache<T>(
+    private val maxSize: Int,
+    private val fetch: (BoundingBox) -> Map<T, LatLon>,
+    private val onEntriesRemoved: (Set<T>) -> Unit
+) {
+    private val byTile = LinkedHashMap<TilePos, HashSet<T>>((maxSize/0.75).toInt(), 0.75f, true)
+    private val byKey = hashMapOf<T, LatLon>()
+
+    fun remove(key: T) {
+        val pos = byKey.remove(key) ?: return
+        byTile[getTilePosFor(pos)]?.remove(key)
+        onEntriesRemoved(setOf(key))
+    }
+
+    fun removeAll(keys: Collection<T>) {
+        val removed = keys.filter {
+            val pos = byKey.remove(it) ?: return@filter false
+            byTile[getTilePosFor(pos)]?.remove(it)
+            true
+        }
+        onEntriesRemoved(removed.toSet())
+    }
+
+    // returns whether the key was put in cache
+    fun putIfTileExists(key: T, position: LatLon): Boolean {
+        val previousPosition = byKey[key]
+        // in cache already
+        if (previousPosition != null) {
+            // but moved -> remove
+            if (previousPosition != position) remove(key)
+            else return true
+        }
+        val tileInCache = byTile[getTilePosFor(position)]
+        if (tileInCache  != null) {
+            tileInCache.add(key)
+            byKey[key] = position
+            return true
+        }
+        return false
+    }
+
+    fun replaceAllInBBox(entries: Map<T, LatLon>, bbox: BoundingBox) {
+        replaceAllInTiles(entries, bbox.asListOfEnclosingTilePos())
+        // trim only when bbox!
+        trim()
+    }
+
+    // may add tiles, but does not call trim()
+    private fun replaceAllInTiles(entries: Map<T, LatLon>, tiles: Collection<TilePos>) {
+        // create / replace tiles
+        tiles.forEach { tile ->
+            byTile.remove(tile)?.let { byKey.keys.removeAll(it) }
+            byTile[tile] = hashSetOf()
+        }
+        if (tiles.size == 1) {
+            // shortcut: replace old tile
+            byTile[tiles.single()] = entries.keys.toHashSet()
+            byKey.putAll(entries)
+        } else {
+            // put only what is inside tiles
+            tiles.forEach { tile ->
+                val bbox = tile.asBoundingBox(TILE_ZOOM)
+                entries.forEach { (key, pos) ->
+                    if (bbox.contains(pos)) {
+                        byTile[tile]!!.add(key)
+                        byKey[key] = pos
+                    }
+                }
+            }
```

Comment:
This could be optimized due to the fact that any given `pos` can only be in one `tile`/`bbox` as those do not intersect. Maybe like this:

```suggestion
            entries.forEach { (key, pos) ->
                byTile[pos.enclosingTilePos(TILE_ZOOM)]!!.add(key)
                byKey[key] = pos
            }
```